examples/timer_functional/src/main.rs

```diff
@@ -0,0 +1,166 @@
+use std::rc::Rc;
+
+use gloo::timers::callback::{Interval, Timeout};
+use yew::prelude::*;
+
+fn get_current_time() -> String {
+    let date = js_sys::Date::new_0();
+    String::from(date.to_locale_time_string("en-US"))
+}
+
+enum TimerAction {
+    Add(&'static str),
+    Cancel,
+    SetInterval(Interval),
+    SetTimeout(Timeout),
+    TimeoutDone,
+}
+
+#[derive(Clone, Debug)]
+struct TimerState {
+    messages: Vec<&'static str>,
+    interval_handle: Option<Rc<Interval>>,
+    timeout_handle: Option<Rc<Timeout>>,
+}
+
+impl PartialEq for TimerState {
+    fn eq(&self, other: &Self) -> bool {
+        self.messages == other.messages
+            && self.interval_handle.is_some() == other.interval_handle.is_some()
+    }
+}
+
+impl Reducible for TimerState {
+    type Action = TimerAction;
+
+    fn reduce(self: Rc<Self>, action: TimerAction) -> Rc<Self> {
+        match action {
+            TimerAction::Add(message) => {
+                let mut messages = self.messages.clone();
+                messages.push(message);
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: self.interval_handle.clone(),
+                    timeout_handle: self.timeout_handle.clone(),
+                })
+            }
+            TimerAction::SetInterval(t) => Rc::new(TimerState {
+                messages: vec!["Interval started!"],
+                interval_handle: Some(Rc::from(t)),
+                timeout_handle: self.timeout_handle.clone(),
+            }),
+            TimerAction::SetTimeout(t) => Rc::new(TimerState {
+                messages: vec!["Timer started!!"],
+                interval_handle: self.interval_handle.clone(),
+                timeout_handle: Some(Rc::from(t)),
+            }),
+            TimerAction::TimeoutDone => {
+                let mut messages = self.messages.clone();
+                messages.push("Done!");
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: self.interval_handle.clone(),
+                    timeout_handle: None,
+                })
+            }
+            TimerAction::Cancel => {
+                let mut messages = self.messages.clone();
+                messages.push("Canceled!");
+                Rc::new(TimerState {
+                    messages,
+                    interval_handle: None,
+                    timeout_handle: self.timeout_handle.clone(),
+                })
+            }
```

Comment:
The `Cancel` button is displayed as clickable when the `Start Timeout` has been activated and the timeout is running, yet when the `Cancel` button is clicked in this state the interface outputs `Canceled!` and does not stop the timeout.

I think that the `Cancel` button should either not be clickable while the timeout is taking place, or that `timeout_handle` should be set as `None` when the `Cancel` button is clicked so that the timeout is stopped.

It would make more sense though to take the second option to give `Cancel` more functionality I believe:
```suggestion
            TimerAction::Cancel => {
                let mut messages = self.messages.clone();
                messages.push("Canceled!");
                Rc::new(TimerState {
                    messages,
                    interval_handle: None,
                    timeout_handle: None,
                })
            }
```